mops/src/hpe/cpu.tpp

```diff
@@ -2,105 +2,282 @@
+                    std::array<scalar_t, simd_element_count> temp;
+                    for (size_t l = 0; l < simd_element_count; l++) temp[l] = base_multiplier[l];
+                    for (uint8_t j_factor = 0; j_factor < polynomial_order; j_factor++) {
+                        if (j_factor == i_factor) continue;
+                        scalar_t* a_ptr_i_j_factor = a_ptr_i + indices_a_ptr_j[j_factor] * simd_element_count;
+                        for (size_t l = 0; l < simd_element_count; l++) temp[l] *= a_ptr_i_j_factor[l];
+                    }
+                    scalar_t* grad_a_ptr_i_i_factor = grad_a_ptr_i + indices_a_ptr_j[i_factor] * simd_element_count;
+                    for (size_t l = 0; l < simd_element_count; l++) grad_a_ptr_i_i_factor[l] += temp[l];
+                }
+                indices_a_ptr_j += polynomial_order;
+            }
+        }
+
+        grad_o_ptr += size_first_dimension_interleft*simd_element_count;  // shift grad_o to the remainder values
+        a_ptr += size_first_dimension_interleft*simd_element_count*n_possible_factors;  // shift A to the remainder values
+        for (size_t i = 0; i < size_remainder; i++) {
             scalar_t grad_output_i = grad_o_ptr[i];
             size_t i_shift = i * n_possible_factors;
             scalar_t* a_ptr_i = a_ptr + i_shift;
-            scalar_t* grad_A_ptr_i = grad_A_ptr + i_shift;
-            int32_t* indices_ptr_j = indices_A_ptr;
+            scalar_t* grad_a_ptr_i = remainder_grad_a_ptr + i_shift;
+            int32_t* indices_a_ptr_j = indices_a_ptr;
             for (size_t j = 0; j < n_monomials; j++) {
                 scalar_t base_multiplier = grad_output_i*c_ptr[j];
                 for (uint8_t i_factor = 0; i_factor < polynomial_order; i_factor++) {
                     scalar_t temp = base_multiplier;
                     for (uint8_t j_factor = 0; j_factor < polynomial_order; j_factor++) {
                         if (j_factor == i_factor) continue;
-                        temp *= a_ptr_i[indices_ptr_j[j_factor]];
+                        temp *= a_ptr_i[indices_a_ptr_j[j_factor]];
                     }
-                    grad_A_ptr_i[indices_ptr_j[i_factor]] += temp;
+                    grad_a_ptr_i[indices_a_ptr_j[i_factor]] += temp;
                 }
-                indices_ptr_j += polynomial_order;
+                indices_a_ptr_j += polynomial_order;
             }
         }
 
+        un_interleave_tensor<scalar_t, simd_element_count>(grad_A, interleft_grad_a_ptr, remainder_grad_a_ptr);
+
+        delete[] interleft_a_ptr;
+        delete[] remainder_a_ptr;
+        delete[] interleft_grad_a_ptr;
+        delete[] remainder_grad_a_ptr;
+    }
+}
+
+
+template<typename scalar_t>
+void mops::homogeneous_polynomial_evaluation_vjp(
+    Tensor<scalar_t, 2> grad_A,
+    Tensor<scalar_t, 1> grad_output,
+    Tensor<scalar_t, 2> A,
+    Tensor<scalar_t, 1> C,
+    Tensor<int32_t, 2> indices_A
+) {
+    // TODO: checks
```

Comment:
Still TODO?

Reply:
Yes, you're right. The checks are now being done